src/librustc_typeck/astconv.rs

```diff
@@ -1487,6 +1488,22 @@ impl<'o, 'gcx: 'tcx, 'tcx> AstConv<'gcx, 'tcx>+'o {
                                     did,
                                     base_segments.last().unwrap())
             }
+            Def::Variant(did) if permit_variants => {
+                // Convert "variant type" as if it were a real type.
+                // The resulting `Ty` is type of the variant's enum for now.
+                tcx.prohibit_type_params(base_segments.split_last().unwrap().1);
+                let mut ty = self.ast_path_to_ty(rscope,
+                                                 span,
+                                                 param_mode,
+                                                 tcx.parent_def_id(did).unwrap(),
+                                                 base_segments.last().unwrap());
+                if ty.is_fn() {
+                    // Tuple variants have fn type even in type namespace,
+                    // extract true variant type from it.
+                    ty = tcx.no_late_bound_regions(&ty.fn_ret()).unwrap();
```

Comment:
No, due to `tcx.parent_def_id(did)` a few lines above.
Only type arguments are taken from the variant (from `base_segments.last()`), `ast_path_to_ty` works with the enum after that and returns enum's type, which cannot be a function.